components/places/src/storage/history.rs

```diff
@@ -1144,6 +1144,35 @@ pub fn get_visit_page(
     Ok(HistoryVisitInfos { infos })
 }
 
+pub fn get_visit_page_with_bound(
+    db: &PlacesDb,
+    offset: i64,
+    bound: i64,
+    count: i64,
+    exclude_types: VisitTransitionSet,
+) -> Result<HistoryVisitInfos> {
+    let allowed_types = exclude_types.complement();
+    let infos = db.query_rows_and_then_named_cached(
+        "SELECT h.url, h.title, v.visit_date, v.visit_type, h.hidden
+         FROM moz_places h
+         JOIN moz_historyvisits v
+           ON h.id = v.place_id
+         WHERE ((1 << v.visit_type) & :allowed_types) != 0 AND
+               NOT h.hidden AND v.visit_date <= :bound
```

Comment:
nit: could you please put the `AND v.visit_date <= :bound` on its own line; it's pretty important to the query so making it more visually distinct seems helpful.

Reply:
Done.